src/Build.UnitTests/Parser_Tests.cs

```diff
@@ -527,5 +527,29 @@ public void VerifyNoWarningForOrder()
             // Make sure the log contains the correct strings.
             Assert.DoesNotContain("MSB4130:", ml.FullLog); // "No need to warn for this expression - ($(a) == 1 or $(b) == 2) and $(c) == 3."
         }
+
+        // see https://github.com/dotnet/msbuild/issues/5436
+        [Fact]
+        public void SupportItemDefinationGroupInWhenOtherwise()
+        {
+            MockLogger ml = ObjectModelHelpers.BuildProjectExpectSuccess(@"
+                    <Project ToolsVersion=`msbuilddefaulttoolsversion` xmlns=`msbuildnamespace`>
+                        <Choose>
+                            <When Condition=` '$(OutputType)'=='Library' `>
```

Comment:
I would expect tests that
-verify that the IDG works, and isn't ignored 
-verify that multiple IDG are supported and they are processed top to bottom
-this for multiple When, as well as Otherwise

Reply:
I think testing the same way as `ItemGroup` might cover all the scenarios. That could be a start.

Reply:
Also put in some PropertyGroup and ItemGroup in there. Everything is evaluated in passes -- normally, all properties, then all IDG, then all items. each pass can "see" the previous, even if the elements were "lower" in the project file. So if there are PG and IG in here, whatever the ordering, the IG should see the IDG.

This comment is now wrong:
https://github.com/dotnet/msbuild/blob/5d102ae37163fc1f22610bb433c3ab9a2fff91f0/src/Build/Evaluation/Evaluator.cs#L1495
it's certainly important to verify that the IDG are "working" inside Choose.

Reply:
@danmoseley  Should we add PropertyGroup and ItemGroup in the project definition, and verify the itemdefinitiongroup works there in the test?

Reply:
> /// We enter here in both the property and item passes, since Chooses can contain both.

Will remove this line or update to "We enter here in the property, itemDefinition, and item passes, since Chooses can contain all of them" ?

Reply:
> @danmoseley  Should we add PropertyGroup and ItemGroup in the project definition, and verify the itemdefinitiongroup works there in the test?  

Sorry I'm not sure what you mean. Yes, I would expect tests that verify the IDG work, including consuming property expressions

Reply:
There's some confusion here, so I'll just lay out roughly what I'd want the test to look like:
Project
Choose
When
ItemGroup
ItemDefinitionGroup
PropertyGroup
Otherwise
ItemDefinitionGroup
PropertyGroup
ItemGroup

For each ItemDefinitionGroup, PropertyGroup, and ItemGroup, there should be things actually defined in it, so like for a PropertyGroup:
<PropertyGroup>
<Foo>bar</Foo>
</PropertyGroup

That's important because you can have something that parses without throwing an error but doesn't actually work, and you want to verify that it works.

So then you also want to make sure each of the Property/Item/ItemDefinitionGroups that you'd expected to execute actually did execute. It would be good to do that for both the When and the Otherwise case, so a Theory with one case for each would be appropriate here.

Is that clearer? (And I'm guessing that hits what you wanted, right danmoseley?)

Reply:
@danmoseley   I have updated the test case, please have a look. Any suggestions, please let me know. Thank you.

Reply:
This looks much closer to what I'd expect, but the When and Otherwise conditions are currently identical. That means we don't know if it's following the When, Otherwise, or both. It would be good to disentangle those.

Reply:
Add one condition to distinguish when it is following when or otherwise